README.md
# safeguard-sessions-plugin-hashicorp-vault
HashiCorp Vault plugin for Safeguard for Privileged Sessions

# Support
- This plugin is officially supported for One Identity Safeguard for Privileged Sessions (SPS) customers with a valid maintenance contract. Feature requests and bug reports should be submitted via the regular support channel on the [One Identity Support page](https://support.oneidentity.com). Such requests are tracked and will be answered within the time frames set out in the support agreement.

- We welcome feature requests, pull requests, and bug reports submitted via GitHub.com, but in this case there is no guaranteed response time.

# About us

One Identity solutions eliminate the complexities and time-consuming processes often required to govern identities, manage privileged accounts and control access. Our solutions enhance business agility while addressing your IAM challenges with on-premises, cloud and hybrid environments.

# Contacting us

For sales or other inquiries, visit the [One Identity Contact page](https://www.oneidentity.com/company/contact-us.aspx) or call +1-800-306-9329.

# Technical support resources

Technical support is available to One Identity customers with a valid maintenance contract and customers who have trial versions. You can access the Support Portal at the [One Identity Support page](https://support.oneidentity.com/).

The Support Portal provides self-help tools you can use to solve problems quickly and independently, 24 hours a day, 365 days a year. The Support Portal enables you to:

- Submit and manage a Service Request
- View Knowledge Base articles
- Sign up for product notifications
- Download software and technical documentation
- View how-to-videos at [YouTube](https://www.YouTube.com/OneIdentity)
- Engage in community discussions
- Chat with support engineers online
- View services to assist you with your product


CHANGELOG
2.0.0 2020-08-05

Use new credential store base class features to make plugin code easier to maintain

1.1.2 2019-11-15

Give use_credential parameter a default value, as written in the default.cfg

1.1.1 2019-10-07

INCLUDES INCOMPATIBLE CONFIG CHANGE!
Rewrite default.cfg to be more logical and also according to documentation guidelines.
Move password_field, key_field, delimiter under secret-kv-v1 to be more readable.

1.1.0 2019-10-02

Add LDAP and userpass authentication methods.
Make it possible to retrieve private keys.
Rename `hashicorp-vault` section to `hashicorp` in config.
Let end-user define the secret path as an answer to an interactive question in an AA plugin.
Make password and private key field name configurable and also let the end-user specify it as part of the secret path.

1.0.0 2019-06-11

Initial release on GitHub.


lib/tests/conftest.py
#
#   Copyright (c) 2019 One Identity
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
#
from textwrap import dedent
import pytest


@pytest.fixture
def hc_address(site_parameters):
    return site_parameters["address"]


@pytest.fixture
def hc_port(site_parameters):
    return site_parameters["port"]


@pytest.fixture
def hc_role(site_parameters):
    return site_parameters["role"]


@pytest.fixture
def hc_vault_token(site_parameters):
    return site_parameters["vault_token"]


@pytest.fixture
def hc_secret_path(site_parameters):
    return site_parameters["secrets_path"]


@pytest.fixture
def hc_account(site_parameters):
    return site_parameters["account"]


@pytest.fixture
def hc_account_password(site_parameters):
    return site_parameters["account_password"]


@pytest.fixture
def hc_wrong_account(site_parameters):
    return site_parameters["wrong_account"]


@pytest.fixture
def hc_account_with_private_key(site_parameters):
    return site_parameters["account_with_private_key"]


@pytest.fixture()
def hc_account_private_key(site_parameters):
    return site_parameters["account_private_key"]


@pytest.fixture()
def hc_account_with_unsupported_key(site_parameters):
    return site_parameters["account_with_unsupported_key"]


@pytest.fixture
def hc_config_approle_auth_engine_kv_v1(site_parameters):
    yield dedent(
        """
        [hashicorp]
        address = {address}
        port = {port}
        authentication_method = approle

        [approle-authentication]
        role = {role}
        vault_token = {vault_token}

        [engine-kv-v1]
        secrets_path = {secrets_path}
    """.format(
            address=site_parameters["address"],
            port=site_parameters["port"],
            role=site_parameters["role"],
            vault_token=site_parameters["vault_token"],
            secrets_path=site_parameters["secrets_path"],
        )
    )


@pytest.fixture
def make_hc_config(site_parameters):
    def _make_config(auth_method, secrets_path=site_parameters["secrets_path"], extra_conf=""):
        username = password = None
        if auth_method == "ldap":
            username = site_parameters["ldap_username"]
            password = site_parameters["ldap_password"]
        elif auth_method == "userpass":
            username = site_parameters["username"]
            password = site_parameters["password"]

        return dedent(
            """
            [logging]
            log_level=debug

            [hashicorp]
            address = {address}
            port = {port}
            authentication_method = {auth_method}
            use_credential=explicit
            username={username}
            password={password}

            [approle-authentication]
            role = {role}
            vault_token = {vault_token}

            [engine-kv-v1]
            secrets_path = {secrets_path}
            {extra_conf}
        """.format(
                address=site_parameters["address"],
                port=site_parameters["port"],
                role=site_parameters["role"],
                vault_token=site_parameters["vault_token"],
                secrets_path=secrets_path,
                username=username,
                password=password,
                auth_method=auth_method,
                extra_conf=extra_conf,
            )
        )

    return _make_config


@pytest.fixture
def generate_params():
    def _params(cookie={}, session_cookie={}, server_username=None, protocol=None):
        return {
            "cookie": cookie,
            "session_cookie": session_cookie,
            "session_id": "example-1",
            "client_ip": "1.1.1.1",
            "client_hostname": None,
            "gateway_domain": None,
            "gateway_username": "wsmith",
            "gateway_password": "",
            "gateway_groups": None,
            "server_username": server_username,
            "server_ip": "1.2.3.4",
            "server_port": 22,
            "server_hostname": None,
            "server_domain": None,
            "protocol": protocol,
        }
    return _params


lib/tests/test_client.py
#
#   Copyright (c) 2019 One Identity
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
#
import pytest
import json
from requests.exceptions import ConnectionError
from pytest import raises
from unittest.mock import patch, call, MagicMock
from collections import namedtuple
from textwrap import dedent
from contextlib import contextmanager

from safeguard.sessions.plugin.logging import configure as log_configure
from safeguard.sessions.plugin.plugin_configuration import PluginConfiguration
from safeguard.sessions.plugin.endpoint_extractor import EndpointException

from ..client import (
    Client,
    InvalidConfigurationError,
    VaultException,
    AppRoleAuthenticator,
    PasswordTypeAuthenticator,
    KVEngineV1SecretRetriever,
)

Response = namedtuple("Response", "ok text status_code")

ADDRESS = "vault-address"
PORT = 1337
URL = "http://{}:{}".format(ADDRESS, PORT)
ROLE = "test-role"
SECRET_KEY = "key"
SECRET = "secret"
SECRETS_PATH = "kv/secret"
ROLE_ID = "9eef5cf1-19d3-0cfb-b9bb-8b6248a17ece"
SECRET_ID = "570bb5c9-2083-242d-5ba0-71224c1d4c48"
CLIENT_TOKEN = "s.menL5xntpcw7RY2fMGDVU4Bo"
VAULT_TOKEN = "s.AjtEPrapWWagEqXFtoNPOpaf"
AUTHENTICATOR = AppRoleAuthenticator(URL, VAULT_TOKEN, ROLE)
SECRET_RETRIEVER = KVEngineV1SecretRetriever(URL, SECRETS_PATH)

ROLE_ID_ENDPOINT = URL + "/v1/auth/approle/role/" + ROLE + "/role-id"
SECRET_ID_ENDPOINT = URL + "/v1/auth/approle/role/" + ROLE + "/secret-id"
LOGIN_ENDPOINT = URL + "/v1/auth/approle/login"
SECRETS_ENDPOINT = URL + "/v1/" + SECRETS_PATH


HASHICORP_VAULT_APPROLE_AUTH_CONFIG = dedent(
    """
    [approle-authentication]
    role = {}
    vault_token = {}
""".format(
        ROLE, VAULT_TOKEN
    )
)

HASHICORP_VAULT_KV_V1_CONFIG = dedent(
    """
    [engine-kv-v1]
    secrets_path = {}
""".format(
        SECRETS_PATH
    )
)


POST_RESPONSES = [
    Response(
        text=json.dumps(
            {
                "request_id": "5dff16d1-6faa-9a3d-c1d7-0cfe4528701b",
                "lease_id": "",
                "renewable": False,
                "lease_duration": 0,
                "data": {"secret_id": SECRET_ID, "secret_id_accessor": "97100e76-87e6-a0ea-a9eb-68388cc0ec0d"},
                "wrap_info": None,
                "warnings": None,
                "auth": None,
            }
        ),
        ok=True,
        status_code=200,
    ),
    Response(
        text=json.dumps(
            {
                "request_id": "c0047be1-ce15-e7cd-3abb-bdba02888783",
                "lease_id": "",
                "renewable": False,
                "lease_duration": 0,
                "data": None,
                "wrap_info": None,
                "warnings": None,
                "auth": {
                    "client_token": CLIENT_TOKEN,
                    "accessor": "DeI3XGrWkjYBscmrTJI2BthN",
                    "policies": ["default", "dev-policy", "my-policy"],
                    "token_policies": ["default", "dev-policy", "my-policy"],
                    "metadata": {"role_name": ROLE},
                    "lease_duration": 2764800,
                    "renewable": True,
                    "entity_id": "39a045eb-8556-c30c-05f1-2be0525080b4",
                    "token_type": "service",
                    "orphan": True,
                },
            }
        ),
        ok=True,
        status_code=200,
    ),
]

GET_RESPONSES = [
    Response(
        text=json.dumps(
            {
                "initialized": True,
                "sealed": False,
                "standby": False,
                "performance_standby": False,
                "replication_performance_mode": "disabled",
                "replication_dr_mode": "disabled",
                "server_time_utc": 1568365137,
                "version": "1.1.2",
                "cluster_name": "vault-cluster-eba4a26a",
                "cluster_id": "df107efd-43e2-b4a0-a0ac-939e3cead978",
            }
        ),
        ok=True,
        status_code=200,
    ),
    Response(
        text=json.dumps(
            {
                "request_id": "39dcab3e-6890-fbf4-460b-d3f1f49df2b3",
                "lease_id": "",
                "renewable": False,
                "lease_duration": 0,
                "data": {"role_id": ROLE_ID},
                "wrap_info": None,
                "warnings": None,
                "auth": None,
            }
        ),
        ok=True,
        status_code=200,
    ),
    Response(
        text=json.dumps(
            {
                "request_id": "969945d0-e8b2-f154-8ca9-9ebca3c25b1e",
                "lease_id": "",
                "renewable": False,
                "lease_duration": 2764800,
                "data": {SECRET_KEY: SECRET},
                "wrap_info": None,
                "warnings": None,
                "auth": None,
            }
        ),
        ok=True,
        status_code=200,
    ),
]

ERROR_RESPONSE_NOT_FOUND = Response(text=json.dumps({"errors": ["This is a generic error"]}), ok=False, status_code=404)


ERROR_RESPONSE_NO_SUCH_FIELD = Response(
    text=json.dumps(
        {
            "request_id": "969945d0-e8b2-f154-8ca9-9ebca3c25b1e",
            "lease_id": "",
            "renewable": False,
            "lease_duration": 2764800,
            "data": {},
            "wrap_info": None,
            "warnings": None,
            "auth": None,
        }
    ),
    ok=True,
    status_code=200,
)


PERFORMANCE_STANDBY_HEALTH_RESPONSE = Response(
    text=json.dumps(
        {
            "initialized": True,
            "sealed": False,
            "standby": False,
            "performance_standby": True,
            "replication_performance_mode": "disabled",
            "replication_dr_mode": "disabled",
            "server_time_utc": 1568365137,
            "version": "1.1.2",
            "cluster_name": "vault-cluster-eba4a26a",
            "cluster_id": "df107efd-43e2-b4a0-a0ac-939e3cead978",
        }
    ),
    ok=True,
    status_code=473,
)

VAULT_DOWN = Response(
    text=json.dumps(
        {
            "initialized": True,
            "sealed": False,
            "standby": False,
            "performance_standby": False,
            "replication_performance_mode": "disabled",
            "replication_dr_mode": "disabled",
            "server_time_utc": 1568365137,
            "version": "1.1.2",
            "cluster_name": "vault-cluster-eba4a26a",
            "cluster_id": "df107efd-43e2-b4a0-a0ac-939e3cead978",
        }
    ),
    ok=False,
    status_code=401,
)


@contextmanager
def _open_session():
    yield SESSION


SESSION = MagicMock()
REQUESTS_TLS = MagicMock()
REQUESTS_TLS.tls_enabled = False
REQUESTS_TLS.open_session = _open_session


def hashicorp_vault_config(address=ADDRESS, auth_method="approle", extra_parts=""):
    return dedent(
        """
        [logging]
        log_level=debug

        [hashicorp]
        address = {}
        port = {}
        authentication_method = {}
        {}
        """.format(
            address, PORT, auth_method, extra_parts
        )
    )


def test_client_can_be_instantiated():
    client = Client(REQUESTS_TLS, AUTHENTICATOR, SECRET_RETRIEVER)
    assert isinstance(client, Client)


@patch("safeguard.sessions.plugin.requests_tls.RequestsTLS", return_value=REQUESTS_TLS)
def test_client_can_be_instantiated_with_config(_requests_tls, mocker):
    SESSION.post.side_effect = POST_RESPONSES
    SESSION.get.side_effect = GET_RESPONSES
    config = create_config_and_set_loglevel(
        hashicorp_vault_config() + HASHICORP_VAULT_APPROLE_AUTH_CONFIG + HASHICORP_VAULT_KV_V1_CONFIG
    )
    mocker.spy(Client, "__init__")
    mocker.spy(AppRoleAuthenticator, "__init__")
    mocker.spy(KVEngineV1SecretRetriever, "__init__")
    client = Client.create_client(config, "username", "pw")
    client.__init__.assert_called_with(client, client.session_factory, client.authenticator, client.secret_retriever)
    client.authenticator.__init__.assert_called_with(client.authenticator, URL, VAULT_TOKEN, ROLE)
    client.secret_retriever.__init__.assert_called_with(client.secret_retriever, URL, SECRETS_PATH)


@patch("safeguard.sessions.plugin.requests_tls.RequestsTLS", return_value=REQUESTS_TLS)
def test_client_factory_uses_HTTPS_when_TLS_enabled(_requests_tls, mocker):
    SESSION.get.side_effect = GET_RESPONSES
    REQUESTS_TLS.tls_enabled = True
    https_url = "https://{}:{}".format(ADDRESS, PORT)
    config = create_config_and_set_loglevel(
        hashicorp_vault_config() + HASHICORP_VAULT_APPROLE_AUTH_CONFIG + HASHICORP_VAULT_KV_V1_CONFIG
    )
    mocker.spy(AppRoleAuthenticator, "__init__")
    mocker.spy(KVEngineV1SecretRetriever, "__init__")
    client = Client.create_client(config, "myuser", "mypassword")
    client.authenticator.__init__.assert_called_with(client.authenticator, https_url, VAULT_TOKEN, ROLE)
    client.secret_retriever.__init__.assert_called_with(client.secret_retriever, https_url, SECRETS_PATH)


@patch("safeguard.sessions.plugin.requests_tls.RequestsTLS", return_value=REQUESTS_TLS)
def test_client_factory_raises_exception_if_secrets_engine_cannot_be_determined(_requests_tls):
    SESSION.get.side_effect = GET_RESPONSES
    config = create_config_and_set_loglevel(hashicorp_vault_config() + HASHICORP_VAULT_APPROLE_AUTH_CONFIG)
    with raises(InvalidConfigurationError):
        Client.create_client(config, "myuser", "mypassword")


def test_get_secret_by_key():
    SESSION.post.side_effect = POST_RESPONSES
    SESSION.get.side_effect = GET_RESPONSES[1:]
    expected_headers = {"X-Vault-Token": VAULT_TOKEN}
    expected_calls_to_get = [
        call(ROLE_ID_ENDPOINT, headers=expected_headers, params=None),
        call(SECRETS_ENDPOINT, headers={"X-Vault-Token": CLIENT_TOKEN}, params=None),
    ]
    expected_calls_to_post = [
        call(SECRET_ID_ENDPOINT, headers=expected_headers, data=None, params=None),
        call(LOGIN_ENDPOINT, headers=expected_headers, data=json.dumps({"role_id": ROLE_ID, "secret_id": SECRET_ID}), params=None),
    ]

    client = Client(REQUESTS_TLS, AUTHENTICATOR, SECRET_RETRIEVER)
    secret = client.get_secret(key=SECRET_KEY)

    assert secret == SECRET
    SESSION.get.assert_has_calls(calls=expected_calls_to_get, any_order=False)
    SESSION.post.assert_has_calls(calls=expected_calls_to_post, any_order=False)


def test_error_occurs_when_getting_secret():
    SESSION.post.side_effect = POST_RESPONSES
    SESSION.get.side_effect = [GET_RESPONSES[1], ERROR_RESPONSE_NOT_FOUND]
    client = Client(REQUESTS_TLS, AUTHENTICATOR, SECRET_RETRIEVER)
    with raises(EndpointException) as exc:
        client.get_secret(key=SECRET_KEY)
    assert exc.match("Endpoint responded with NOK;")


def test_error_occurs_when_extracting_secret():
    SESSION.post.side_effect = POST_RESPONSES
    SESSION.get.side_effect = [GET_RESPONSES[1], ERROR_RESPONSE_NO_SUCH_FIELD]
    client = Client(REQUESTS_TLS, AUTHENTICATOR, SECRET_RETRIEVER)
    with raises(EndpointException) as exc:
        client.get_secret(key=SECRET_KEY)
    assert exc.match("Endpoint \\(http://vault-address:1337/v1/kv/secret\\) response did not contain the information;")


def test_cannot_connect_to_vault():
    SESSION.get.side_effect = [VAULT_DOWN]
    with raises(VaultException):
        Client._determine_vault_to_use(REQUESTS_TLS, ["wrong.address"], "1234")


def data_provider():
    yield ("ldap", PasswordTypeAuthenticator, "use_credential=explicit\nusername=user\npassword=pass")
    yield ("userpass", PasswordTypeAuthenticator, "use_credential=explicit\nusername=user\npassword=pass")
    yield ("approle", AppRoleAuthenticator, "")


@patch("safeguard.sessions.plugin.requests_tls.RequestsTLS", return_value=REQUESTS_TLS)
@pytest.mark.parametrize("auth_method, instance, extra_parts", data_provider())
def test_client_uses_the_appropriate_authenticator(_requests_tls, auth_method, instance, extra_parts):
    SESSION.get.side_effect = GET_RESPONSES
    config = create_config_and_set_loglevel(
        hashicorp_vault_config(auth_method=auth_method, extra_parts=extra_parts)
        + HASHICORP_VAULT_APPROLE_AUTH_CONFIG
        + HASHICORP_VAULT_KV_V1_CONFIG
    )
    client = Client.create_client(config, "myuser", "mypassword")
    assert isinstance(client.authenticator, instance)
    assert client.authenticator.authentication_backend == auth_method


@patch("safeguard.sessions.plugin.requests_tls.RequestsTLS", return_value=REQUESTS_TLS)
def test_raises_vault_error_when_cannot_make_connection_to_vault(_requests_tls):
    SESSION.get.side_effect = [ConnectionError()]
    config = create_config_and_set_loglevel(
        hashicorp_vault_config(
            address="vault.is.down",
            auth_method="ldap",
            extra_parts="use_credential=explicit\nusername=user\npassword=pass",
        )
        + HASHICORP_VAULT_APPROLE_AUTH_CONFIG
        + HASHICORP_VAULT_KV_V1_CONFIG
    )
    with raises(VaultException):
        Client.create_client(config, "myuser", "mypassword")


@patch("safeguard.sessions.plugin.requests_tls.RequestsTLS", return_value=REQUESTS_TLS)
def test_uses_first_available_vault_address_when_more_configured(_requests_tls, mocker):
    SESSION.get.side_effect = [VAULT_DOWN, GET_RESPONSES[0]]
    config = create_config_and_set_loglevel(
        hashicorp_vault_config(
            address="vault.is.down,test.vault", auth_method="ldap", extra_parts="use_credential=gateway"
        )
        + HASHICORP_VAULT_APPROLE_AUTH_CONFIG
        + HASHICORP_VAULT_KV_V1_CONFIG
    )
    mocker.spy(PasswordTypeAuthenticator, "__init__")
    client = Client.create_client(config, "myuser", "mypassword")
    client.authenticator.__init__.assert_called_with(
        client.authenticator, "https://test.vault:{}".format(PORT), "myuser", "mypassword", "ldap"
    )


@patch("safeguard.sessions.plugin.requests_tls.RequestsTLS", return_value=REQUESTS_TLS)
def test_uses_first_available_vault_address_even_if_performance_standby(_requests_tls, mocker):
    SESSION.get.side_effect = [VAULT_DOWN, PERFORMANCE_STANDBY_HEALTH_RESPONSE]
    config = create_config_and_set_loglevel(
        hashicorp_vault_config(
            address="vault.is.down,test.vault", auth_method="ldap", extra_parts="use_credential=gateway"
        )
        + HASHICORP_VAULT_APPROLE_AUTH_CONFIG
        + HASHICORP_VAULT_KV_V1_CONFIG
    )
    mocker.spy(PasswordTypeAuthenticator, "__init__")
    client = Client.create_client(config, "myuser", "mypassword")
    client.authenticator.__init__.assert_called_with(
        client.authenticator, "https://test.vault:{}".format(PORT), "myuser", "mypassword", "ldap"
    )


def create_config_and_set_loglevel(config_str):
    config = PluginConfiguration(config_str)
    log_configure(config)
    return config


lib/client.py
#
#   Copyright (c) 2019 One Identity
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
#
from http.client import responses
import requests
import json
import abc
from safeguard.sessions.plugin.requests_tls import RequestsTLS
from safeguard.sessions.plugin.logging import get_logger
from safeguard.sessions.plugin import PluginSDKRuntimeError
from safeguard.sessions.plugin.endpoint_extractor import EndpointExtractor, EndpointException


logger = get_logger(__name__)


class VaultException(PluginSDKRuntimeError):
    pass


class InvalidConfigurationError(PluginSDKRuntimeError):
    pass


class Client:
    USABLE_HEALTH_RESPONSE_CODES = {200: "active", 473: "performance standby"}

    def __init__(self, requests_tls, authenticator, secret_retriever):
        self.__requests_tls = requests_tls
        self.__authenticator = authenticator
        self.__secret_retriever = secret_retriever

    @property
    def authenticator(self):
        return self.__authenticator

    @property
    def secret_retriever(self):
        return self.__secret_retriever

    @property
    def session_factory(self):
        return self.__requests_tls

    @staticmethod
    def token_header_from(token):
        return {"X-Vault-Token": token}

    def get_secret(self, key):
        with self.__requests_tls.open_session() as session:
            client_token = self.__authenticator.authenticate(session)
            secret = self.__secret_retriever.retrieve_secret(session, key, client_token)
        return secret

    @classmethod
    def _determine_vault_to_use(cls, requests_tls, vault_addresses, vault_port):
        vault_url = None
        with requests_tls.open_session() as session:
            for vault_address in vault_addresses:
                vault_url = "{}://{}:{}".format(
                    "https" if requests_tls.tls_enabled else "http", vault_address, vault_port
                )
                try:
                    logger.debug("Try to setup connection to Vault on address: {}".format(vault_address))
                    response = EndpointExtractor().invoke_http_method(session, vault_url + "/v1/sys/health", "get")
                    if response.status_code in cls.USABLE_HEALTH_RESPONSE_CODES:
                        logger.info(
                            'Using Vault {}; status="{}"'.format(
                                vault_url, cls.USABLE_HEALTH_RESPONSE_CODES[response.status_code],
                            )
                        )
                        break
                except EndpointException as exc:
                    raise VaultException("Connection error: {}".format(exc))
                except VaultException:
                    logger.error("Cannot connect to vault on the following address: {}".format(vault_url))
                    continue
            else:
                raise VaultException("None of the configured vaults can be reached")
        return vault_url

    @classmethod
    def create_client(cls, config, auth_username, auth_password, secrets_path=None):
        requests_tls = RequestsTLS.from_config(config)
        vault_addresses = list(map(lambda va: va.strip(), config.get("hashicorp", "address", required=True).split(",")))
        vault_port = config.getint("hashicorp", "port", default=8200)

        vault_url = cls._determine_vault_to_use(requests_tls, vault_addresses, vault_port)

        if not secrets_path:
            secrets_path = config.get("engine-kv-v1", "secrets_path")

        if secrets_path:
            secret_retriever = KVEngineV1SecretRetriever(vault_url, secrets_path)
        else:
            raise InvalidConfigurationError("No valid secrets engine can be determined based on config")

        authenticator = AuthenticatorFactory.create_authenticator(config, vault_url, auth_username, auth_password)
        return cls(requests_tls, authenticator, secret_retriever)


class AuthenticatorFactory:
    @classmethod
    def create_authenticator(cls, config, vault_url, auth_username, auth_password):
        auth_method = config.getienum(
            "hashicorp", "authentication_method", ("ldap", "userpass", "approle"), required=True
        )
        logger.debug("Authenticating to vault with method: {}".format(auth_method))
        if auth_method in ("ldap", "userpass"):
            authenticator = PasswordTypeAuthenticator(vault_url, auth_username, auth_password, auth_method)
        elif auth_method == "approle":
            role = config.get("approle-authentication", "role", required=True)
            vault_token = config.get("approle-authentication", "vault_token")
            authenticator = AppRoleAuthenticator(vault_url, vault_token, role)
        else:
            raise InvalidConfigurationError("No valid authentication method can be determined based on config")
        return authenticator


class SecretRetriever(abc.ABC):
    @abc.abstractmethod
    def retrieve_secret(self, session, key, client_token):
        raise NotImplementedError


class KVEngineV1SecretRetriever(SecretRetriever):
    def __init__(self, vault_url, secrets_path):
        self.__vault_url = vault_url
        self.__secrets_path = secrets_path

    def retrieve_secret(self, session, key, client_token):
        secret = EndpointExtractor().extract_data_from_endpoint(
            session,
            endpoint_url=self.__vault_url + "/v1/" + self.__secrets_path,
            data_path="data." + key,
            headers=Client.token_header_from(client_token),
            method="get",
        )
        return secret


class Authenticator(abc.ABC):
    @abc.abstractmethod
    def authentication_backend(self):
        raise NotImplementedError

    @abc.abstractmethod
    def authenticate(self, session):
        raise NotImplementedError


class AppRoleAuthenticator(Authenticator):
    def __init__(self, vault_url, vault_token, role):
        self.__vault_url = vault_url
        self.__vault_token = vault_token
        self.__role = role

    @property
    def authentication_backend(self):
        return "approle"

    def authenticate(self, session):
        endpoint_extractor = EndpointExtractor(self.__vault_url)
        role_id = endpoint_extractor.extract_data_from_endpoint(
            session,
            endpoint_url="/v1/auth/approle/role/" + self.__role + "/role-id",
            data_path="data.role_id",
            headers=Client.token_header_from(self.__vault_token),
            method="get",
        )
        secret_id = endpoint_extractor.extract_data_from_endpoint(
            session,
            endpoint_url="/v1/auth/approle/role/" + self.__role + "/secret-id",
            data_path="data.secret_id",
            headers=Client.token_header_from(self.__vault_token),
            method="post",
        )
        client_token = endpoint_extractor.extract_data_from_endpoint(
            session,
            endpoint_url="/v1/auth/approle/login",
            data_path="auth.client_token",
            headers=Client.token_header_from(self.__vault_token),
            method="post",
            data={"role_id": role_id, "secret_id": secret_id},
        )
        return client_token


class PasswordTypeAuthenticator(Authenticator):
    def __init__(self, vault_url, username, password, auth_backend):
        self.__vault_url = vault_url
        self.__username = username
        self.__password = password
        self.__auth_backend = auth_backend

    @property
    def authentication_backend(self):
        return self.__auth_backend

    def __calculate_endpoint(self):
        return self.__vault_url + "/v1/auth/" + self.__auth_backend + "/login/" + self.__username

    def authenticate(self, session):
        logger.debug("Performing Userpass authentication for user {}".format(self.__username))
        endpoint = self.__calculate_endpoint()
        client_token = EndpointExtractor().extract_data_from_endpoint(
            session,
            endpoint_url=endpoint,
            data_path="auth.client_token",
            method="post",
            data={"password": self.__password},
        )
        return client_token
